### PR TITLE
NAS oneshot (stage 2) - Supernet modules

### PR DESCRIPTION
**Pull request by ultmaster**

### CODE REVIEW

**Review by QuanluZhang — APPROVED** [review]
it would be better to provide simple tutorials about how to use these operators, sampling, and differentiables.

**Comment by ultmaster** [comment]
> it would be better to provide simple tutorials about how to use these operators, sampling, and differentiables.

Sure. Will do.

File: nni/nas/oneshot/pytorch/supermodule/sampling.py [hunk 1]
```diff
@@ -420,10 +376,16 @@ def mutate(cls, module, name, memo, mutate_kwargs):
...
-                op_factory = {  # create a factory
-                    name: lambda _, __, ___: copy.deepcopy(op_candidates_lc[name])
-                    for name in op_candidates_lc.names
-                }
+                candidates = op_candidates_lc.candidates
+                def _copy(_, __, ___, op):
+                    return copy.deepcopy(op)
+
+                if isinstance(candidates, list):
+                    op_factory = [functools.partial(_copy, op=op) for op in candidates]
```
Comment (liuzhe-lz) [inline nni/nas/oneshot/pytorch/supermodule/sampling.py:384]: Is this equal to `lambda *args: copy.deepcopy(op)`?
The mental overhead is massive...

Comment (liuzhe-lz) [reply nni/nas/oneshot/pytorch/supermodule/sampling.py:384]: Oh `op` is the iterator so you need to shallow copy `op` each time before constructing the lambda...

**Review by liuzhe-lz — APPROVED** [review]

File: nni/nas/oneshot/pytorch/supermodule/sampling.py (hunk 1 above)
Comment (matluster) [reply nni/nas/oneshot/pytorch/supermodule/sampling.py:384]: I suppose there is a better way, but I don't think it's worth the effort.

**Pull request merged**